Replace Appveyor windows testing with GHA

jonchurch opened the pull request:
> Edit: Ah, this will take more work than I realized. our current CI uses several cli tools to get its job done. I can make this work, but not in the 15 mins I had when I first opened the PR. I setup nvm-windows to allow us to install node-versions below 6 (which is lowest supported by actions/node-setup as far as I can tell, which would be my preference) but there is still work to be done here. Maybe we want to use [`actions/setup-node-using-nvm`](https://github.com/marketplace/actions/set-up-node-using-nvm)

Currently we test against `unbuntu-latest` on github actions, and then on windows with appveyor.

I want to remove the appveyor dependency, as it has been hanging and slow, holding up PRs for hours at a time.

This PR adds `windows-latest` to the CI matrix.

I don't know a better way to test this than via PR, and want to run it in the main repo rather than my fork. Will fixup history on my branch once I have something ready to land.

### tasks

- [x] Run tests on windows-latest
- [x] Split the Lint step out so it only runs once, not once per node version/platform combo
- [ ] ~~Drop io.js support officially, testing for it isn't worth it, if we support node 4 we support io.js (I want to do this, but eh, if it's not too much work to continue to run CI for io.js I'll keep it in instead. haven't invested much in this)~~
- [x] Figure out a fix for the `npm install` failure under Node 5
- [ ] Remove Appveyor Connection
- [x] Figure out how we are using codecoverage, where do we report it? (we report it to coveralls, in a CI step, viewable [here](https://coveralls.io/github/expressjs/express))
  - [ ] Investigate reducing coverage runs (do we really need to run for every version?)
- [x] Rename the CI from Experimental
- [x] Replace existing v4 CI
- [ ] Open Issue for v5 CI based on the work here

jonchurch committed:
add windows-latest to the ci matrix

jonchurch committed:
move os into matrix, set runs-on via matrix

jonchurch committed:
add nvm for windows

jonchurch committed:
fixup! move os into matrix, set runs-on via matrix

jonchurch commented:
Note to self: Remove the lint step that runs for every Node Version, it is redundant as we are static analyzing source code, which will not differ based on the Node version.

jonchurch committed:
create cross platform ci, drop linting and code cov for now

jonchurch committed:
fixup! create cross platform ci, drop linting and code cov for now

jonchurch committed:
restore ci.yml to upstream

jonchurch committed:
fixup! create cross platform ci, drop linting and code cov for now

jonchurch committed:
fixup! create cross platform ci, drop linting and code cov for now

jonchurch commented:
Woah, ran into this https://github.com/npm/cli/issues/681 on Node 5 here, dunno what the fix is. I guess update npm? Im just using what ships w/ the latest version of Node 5

jonchurch committed:
add step to disable shrinkwrap/lockfile

jonchurch commented:
Okay, Node 21 seems to have added a new method! `QUERY` which is awesome, looks to have landed here https://github.com/nodejs/node/pull/51719 in `21.7.2` ([changelog](https://github.com/nodejs/node/blob/main/doc/changelogs/CHANGELOG_V21.md#21.7.2))

That's why I have a failing test here on 21, dunno the actual root cause though.

@wesleytodd any idea what needs to be done to router to support QUERY? Or maybe the failure is on Supertest?

To repro it, `nvm install 21 && npm t`

jonchurch committed:
lets go red to make sure things are working

jonchurch committed:
Revert "lets go red to make sure things are working"

This reverts commit 7ae454236ba72ea9f8e165957c2cddf83eda4072.

jonchurch committed:
add 21.7.1 to see if we fail on query there

checking when query came in

jonchurch committed:
Revert "add 21.7.1 to see if we fail on query there"

This reverts commit 4f28818c45e494c4393421398c07c7fef10642b7.

jonchurch committed:
add lint job to experimental ci

jonchurch committed:
name the test job Run tests

jonchurch committed:
add v5 npm install fix

jonchurch committed:
add code coverage upload step

jonchurch committed:
update coverage excludes

jonchurch committed:
fixup! add code coverage upload step

jonchurch committed:
fixup! add v5 npm install fix

jonchurch committed:
fixup! add code coverage upload step

jonchurch committed:
update to setup-node v4

jonchurch committed:
see if setup node will just use io.js for 1-3.x

wesleytodd commented:
> @wesleytodd any idea what needs to be done to router to support QUERY? Or maybe the failure is on Supertest?

Just to close the loop on this, this was sorted out in other channels. It was not yet fixed correctly in core.

jonchurch committed:
add 22.x to matrix

jonchurch committed:
Revert "see if setup node will just use io.js for 1-3.x"

This reverts commit 0fe828f2c0e2b9ca82af3f7c707183f938ac22a0.

jonchurch committed:
collect coverage on every version, merge, upload to coveralls

jonchurch committed:
split iojs CI out from main ci

jonchurch committed:
fix job name conflict

jonchurch committed:
ignore npm-shrinkwrap so we can drop configuring it in CI

jonchurch committed:
remove the shrinkwrap and package-lock configuration in CI

we already set package-lock=false in npmrc, and I just added shrinkwrap
to ignore, we should never encounter it in CI unless we explicitly
create it. I trust us not to do that.

jonchurch committed:
unpin minors

jonchurch committed:
add flag-name to coverage uploads

jonchurch committed:
fix misspelled action name

jonchurch committed:
fixup! unpin minors

jonchurch committed:
add .x to 0.10 and 0.12 to see if setup-node can find those

it's falling back to only downloading the node binary, giving us npm 10
which those version of node cannot run

jonchurch committed:
install npm versions which 0.10 and 0.12 shipped with

jonchurch committed:
install globals if needed, maybe this will work?

jonchurch committed:
fixup! install globals if needed, maybe this will work?

jonchurch committed:
try installing the globals first

jonchurch committed:
fixup! try installing the globals first

jonchurch committed:
fixup! try installing the globals first

jonchurch committed:
further lock the 0.10/.12 npm fix down to windows

jonchurch committed:
fixup! further lock the 0.10/.12 npm fix down to windows

jonchurch committed:
fixup! further lock the 0.10/.12 npm fix down to windows

jonchurch committed:
drop the lts-coverage job

jonchurch committed:
generalize the npm version install

we end up installing npm on non-windows for 0.10 and 0.12 but Im fine
with that to make the steps simpler

jonchurch committed:
fixup! generalize the npm version install

jonchurch committed:
replace ci w/ new ci

jonchurch commented on .gitignore:
```diff
@@ -1,6 +1,7 @@
 # npm
 node_modules
 package-lock.json
+npm-shrinkwrap.json
```
Note that Im adding shrinkwrap to gitignore, we don't generate it, but we did go out of our way in the CI before to configure node to ignore it

Comfortable removing that config part, and just making sure we never commit it

jonchurch commented on package.json:
```diff
@@ -91,8 +91,8 @@
     "test": "mocha --require test/support/env --reporter spec --bail --check-leaks test/ test/acceptance/",
-    "test-ci": "nyc --reporter=lcovonly --reporter=text npm test",
-    "test-cov": "nyc --reporter=html --reporter=text npm test",
+    "test-ci": "nyc --exclude examples --exclude test --exclude benchmarks --reporter=lcovonly --reporter=text npm test",
```
exclude these non source dirs from our coverage report, we don't want to track if these lines of code are exercised through testing.

jonchurch commented on .github/workflows/iojs.yml:
```diff
@@ -0,0 +1,69 @@
+name: iojs-ci
```
This makes cross plat testing so much easier for Node

I switched to `setup-node` to enable cross paltform node setup, but that doesnt' support iojs. Splitting this out here, and only testing iojs on linux makes the most sense to me.

jonchurch committed:
add back in concurrency comment

jonchurch commented on .github/workflows/ci.yml:
```diff
@@ -20,187 +20,149 @@ concurrency:
+        if: ${{matrix.npm-version != ''}}
+        run: npm install -g ${{ matrix.npm-version }}
+
+      - name: Configure npm loglevel
```
dropped configuring npm to ignore package-lock and shrinkwrap

We have .npmrc set to package-lock: false, and this same PR updates gitignore to ignore shrinkwrap

wesleytodd replied:
Is there a reason not to also move this config into the `.npmrc` if this is what we need? I do wonder what the reason was to configure this at all in the past, do you know that?

jonchurch commented on .github/workflows/ci.yml:
```diff
@@ -20,187 +20,149 @@ concurrency:
+        with:
+          node-version: ${{ matrix.node-version }}
+
+      - name: Npm version fixes
```
See the comments in the matrix include section to understand why I added a step to install npm for a few versions

ctcpip commented on .github/workflows/ci.yml:
```diff
@@ -20,187 +20,149 @@ concurrency:
+    steps:
+      - uses: actions/checkout@v4
+      - name: Setup Node.js {{ matrix.node-version }}
+        uses: actions/setup-node@v3
```
```suggestion
        uses: actions/setup-node@v4
```

wesleytodd replied:
Sorry, not seeing where this was resolved. Could we link the convo here on why we are not using the latest for this?

jonchurch replied:
Oops, good catch. Updated to v4 for real for real. Might have force pushed over after taking this in the github ui

blakeembrey reviewed (approved):
LGTM, one nice change would be to remove the dev dependencies that are used only in `examples/` such as `connect-redis` so we don't need to uninstall in tests (add a new `package.json` in the examples instead).

jonchurch commented:
@blakeembrey, it's a good idea. I just did that blindly and then realized that we actually exericse at least some of the examples in our tests.

Leaving this for another day.

We are using these examples in the acceptance:

```
auth
content-negotiation
cookie-sessions
cookies
downloads
ejs
error-pages
error
hello-world
markdown
multi-router
mvc
params
resource
route-map
route-separation
vhost
web-service
```

jonchurch committed:
pin to setup-node@4
